src/ape/api/providers.py

```diff
@@ -1591,6 +1625,283 @@ def _handle_execution_reverted(
+            if start < stop:
+                for block in self.range_blocks(start, stop):
+                    yield block
+
+            if block:
+                last_yielded_hash = block.hash
+                last_yielded_number = block.number
+                time_since_last = time.time()
+            else:
+                _try_timeout()
+
+            time.sleep(block_time)
+
+    def range_contract_events(
+        self,
+        contract: "ContractInstance", # noqa: F821
+        start_or_stop: int,
+        stop: Optional[int] = None,
+        search_topics: Optional[Dict[str, Any]] = None,
+        extra_addresses: Optional[List] = None,
+    ) -> Iterator[ContractLog]:
+        """
+        Search through the logs for this event using the given filter parameters.
+
+        Args:
+            start_or_stop (int): When also given ``stop``, this is the the
+              earliest block number in the desired log set.
+              Otherwise, it is the total amount of blocks to get starting from ``0``.
+            stop (Optional[int]): The latest block number in the
+              desired log set. Defaults to delegating to provider.
+            search_topics (Optional[Dict]): Search topics, such as indexed event inputs,
+              to query by. Defaults to getting all events.
+            extra_addresses (Optional[List[``AddressType``]]): Additional contract
+              addresses containing the same event type. Defaults to only looking at
+              the contract instance where this event is defined.
+
+        Returns:
+            Iterator[:class:`~ape.contracts.base.ContractLog`]
+        """
+
+        if not hasattr(contract, "address"):
+            return
+
+        start_block = None
+        stop_block = None
+
+        if stop is None:
+            contract = None
+            try:
+                contract = self.chain_manager.contracts.instance_at(contract.address)
+            except Exception:
+                pass
+
+            if contract:
+                start_block = contract.receipt.block_number
+            else:
+                start_block = self.chain_manager.contracts.get_creation_receipt(
+                    contract.address
+                ).block_number
+
+            stop_block = start_or_stop
+        elif start_or_stop is not None and stop is not None:
+            start_block = start_or_stop
+            stop_block = stop - 1
+
+        stop_block = min(stop_block, self.chain_manager.blocks.height)
+
+        addresses = set([contract.address] + (extra_addresses or []))
+        contract_event_query = ContractEventQuery(
+            columns=list(ContractLog.__fields__.keys()),
+            contract=addresses,
+            event=self.abi,
+            search_topics=search_topics,
+            start_block=start_block,
+            stop_block=stop_block,
+        )
+        yield from self.query_manager.query(contract_event_query)  # type: ignore
+
+    def poll_logs(
```

Comment:
Technically, for web3py at least, there's a better implementation of this using log filters